certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+
+    // effect
+    assert success => backAfter == value, "back set to value";
+
+    // no side effect
+    assert backBefore == backAfter <=> backBefore == value, "back doesn't change only if it was already set";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: pushBack preserves the previous values in the queue                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBackConsistency {
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 beforeAt = at_(key);
+
+    bytes32 value;
+    pushBack@withrevert(value);
+
+    bytes32 afterAt = at_(key);
+
+    assert afterAt == beforeAt, "data is preserved";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popFront removes an element from the beginning of the queue                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFront {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    bool emptyBefore = empty();
+    bytes32 frontBefore = front@withrevert();
+
+    bytes32 popped = popFront@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => frontBefore == popped, "previous front is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front is previous at(1) after calling popFront                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextFront = at_(1);
+    popFront@withrevert();
+
+    assert !lastReverted <=> front() == nextFront, "front is adjusted";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(x) is preserved and offset to at(x - 1) after calling popFront                                             |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontConsistency {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 before = at_(key);
+
+    popFront@withrevert();
+    bool success = !lastReverted;
+    
+    assert !lastReverted <=> before == at_(key - 1), "values are offset and not modified";
+}
```

Comment:
Suggestion

```suggestion
rule popFrontConsistency(uint256 key) {
    requireInvariant boundariesConsistency();
    require boundedQueue();

    // Read (any) value that is not the front (this assert the value exist / the queue is long enough)
    require at > 1;
    bytes32 before = at_(key);

    // pop front
    popFront();
    
    // try to read value
    bytes32 after = at_@withrevert(key);
    
    assert !lastReverted, "value still exist in queue";
    assert before == after, "values are offset and not modified";
}
```

Reply:
I think you meant `require key > 1;` instead of `require at > 1;` so I added the change with this modification to your suggestion. Let me know if it's not ok.